### DISCUSSION TITLE
ZFS power saving not possible due to spin up every 5 minutes

### DISCUSSION THREAD

**Discussion by runningman84** [siderolabs/talos#8846] [General]
I have installed talos on a new nas node (dxp2800) in my empty test cluster. I have created a raid1 pool instead of the two sata disks:
````
root@worker2:~# zpool list
NAME   SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
tank  10.9T  5.68G  10.9T        -         -     0%     0%  1.00x    ONLINE  -
root@worker2:~# zpool list
NAME   SIZE  ALLOC   FREE  CKPOINT  EXPANDSZ   FRAG    CAP  DEDUP    HEALTH  ALTROOT
tank  10.9T  5.68G  10.9T        -         -     0%     0%  1.00x    ONLINE  -
root@worker2:~# zpool status
  pool: tank
 state: ONLINE
config:

        NAME        STATE     READ WRITE CKSUM
        tank        ONLINE       0     0     0
          mirror-0  ONLINE       0     0     0
            sda     ONLINE       0     0     0
            sdb     ONLINE       0     0     0

root@worker2:~# zfs list
NAME          USED  AVAIL  REFER  MOUNTPOINT
tank         5.68G  10.8T   104K  /var/hddstorage
tank/public  5.68G  10.8T  5.68G  /var/hddstorage/public
````

The spin down setting for both sata drives were set using this command
````
root@worker2:~# hdparm -S 30 /dev/sda

/dev/sda:
 setting standby to 30 (2 minutes + 30 seconds)
root@worker2:~# hdparm -S 30 /dev/sdb

/dev/sdb:
 setting standby to 30 (2 minutes + 30 seconds)
````

With zpool iostat I can monitor the disk activity:
````
root@worker2:~# zpool iostat 1
              capacity     operations     bandwidth 
pool        alloc   free   read  write   read  write
----------  -----  -----  -----  -----  -----  -----
tank        5.68G  10.9T      0      1  4.30K  18.7K
tank        5.68G  10.9T      0      0      0      0
````
This is the full output recording 5 minutes:
https://gist.github.com/runningman84/e5d315b174202b2717594bbc9e34d9b6

The problem is that for some unkown reason every 5 minutes there is some io activty (7 or 8 write iops).

It is unclear to me what is the reason for these operations?

Is this caused by the ext-zpool-importer service? Is there some other talos service which writes to my disks?

**Comment by silpionmatze** [comment]
I also have this problem.

Exactly every 5 minutes

blktrace log:

```
  8,16   9        1     0.000000000  1470  A   W 469779104 + 24 <- (8,17) 469777056
  8,16   9        2     0.000000635  1470  Q   W 469779104 + 24 [z_wr_int_1]
  8,16   9        3     0.000004897  1470  G   W 469779104 + 24 [z_wr_int_1]
  8,16   9        4     0.000005568  1470  P   N [z_wr_int_1]
  8,16   9        5     0.000006317  1470  U   N [z_wr_int_1] 1
  8,16   9        6     0.000007213  1470  I   W 469779104 + 24 [z_wr_int_1]
  8,16   9        7     0.000012109  1470  D   W 469779104 + 24 [z_wr_int_1]
  8,16   0       14     0.000030481  7444  C   W 429716640 + 8 [0]
  8,16   0       15     0.000136773     0  C   W 469779104 + 24 [0]
  8,16  10        1     0.006915126  1568  A   R 44783864 + 8 <- (8,17) 44781816
  8,16  10        2     0.006915370  1568  Q   R 44783864 + 8 [txg_sync]
  8,16  10        3     0.006916296  1568  G   R 44783864 + 8 [txg_sync]
  8,16  10        4     0.006916754  1568  P   N [txg_sync]
  8,16  10        5     0.006916984  1568  U   N [txg_sync] 1
  8,16  10        6     0.006917191  1568  I   R 44783864 + 8 [txg_sync]
  8,16  10        7     0.006918855  1568  D   R 44783864 + 8 [txg_sync]
```

not using talos, so related to zfs `zfs-2.2.4-1`

**Comment by NyCodeGHG** [comment]
Same problem here.
```
zfs-2.2.7-1
zfs-kmod-2.2.7-1
```

@silpionmatze 
@runningman84 
sorry for bumping this old thread, did you find any solution?
I'm completely losing my mind and this is the only post with the same problem I found on the internet.

**Comment by runningman84** [comment]
Unfortunately I haven’t found any solution yet which is very unfortunate. I am now thinking of shutting down the whole storage node instead. In case of usage I can wake it up over lan.

**Comment by p4block** [comment]
Same problem with latest talos 1.12.7 and zfs extension 2.4.1. Massive bummer. I was very excited to move my server/nas to Talos but this is a hard pill to swallow.  

I set spindown time with hdparm on a daemonset but it wasn´t doing anything. I spun down the drives manually with hdparm -Y and yeah, they spin back up for no reason a few mins later. 

These drives are not mounted in the kubelet for pvc provisioning either.

This did not happen on nixos just a few hours ago on same zfs version. There was a bug a while ago, but it got fixed. This version should be clean.

**Comment by p4block** [comment]
I found the problem @runningman84 @NyCodeGHG @silpionmatze it's the pool's `cachefile` setting. zfs tries to write the cachefile to Talos filesystem but fails, and instead writes to the pool itself. Setting `cachefile` to `none` fixes the problem.
I wrote some ramblings about it [here](https://old.reddit.com/r/TalosLinux/comments/1sz8fsp/found_why_talos_zfs_was_waking_my_sleeping_hdds/?)

**Reply by frezbo** [reply]
> A good friend that runs a read only /etc on another distro tells me that they use options zfs spa_config_path= in modprobe.d (empty string works) and problem should be gone. Haven't verified. That has some easy talos equivalent, should just be a matter of adding zfs.sp[...] to kernel cmdline.

i guess setting the module param here would work too:

```yaml
machine:
  kernel:
    modules:
      - name: zfs
        parameters:
          - spa_config_path=
```

would also work

**Reply by frezbo** [reply]
if someone can confirm can update the extension docs

**Reply by runningman84** [reply]
I will try that but would it make sense to configure the cachefile to a local ssd volume which is used in /var?

**Reply by p4block** [reply]
That option seems to be ignored on my system talos 1.13
```
cat /host/sys/module/zfs/parameters/spa_config_path 
/etc/zfs/zpool.cache
```

**Reply by frezbo** [reply]
you'd need a reboot for the module parameters to take effect

**Reply by frezbo** [reply]
> I will try that but would it make sense to configure the cachefile to a local ssd volume which is used in /var?

yes, this would be ideal

**Reply by p4block** [reply]
I did, but it's not affecting it. The talosconfig is applying 100% though. Very weird.

**Reply by frezbo** [reply]
maybe udevd already loaded the module

**Reply by p4block** [reply]
If zfs module is loaded before the extra module load/params, cmdline should suffice, but...

I'm hitting some problem, either in a script somewhere in between keyboard and chair
When I build my image like this `[...]customization":{"extraKernelArgs":["zfs.spa_config_path="],[...]`

the `cat /proc/cmdline` shows it not having the =.
`talos.platform=metal console=tty0 init_on_alloc=1 slab_nomerge pti=on consoleblank=0 nvme_core.io_timeout=4294967295 printk.devkmsg=on selinux=1 module.sig_enforce=1 proc_mem.force_override=never zfs.spa_config_path`
I think that may make it not work, because drives still wake up with this :sweat_smile:

**Reply by frezbo** [reply]
what if you try `zfs.spa_config_path=""`, yeh the kernel module parser will drop `=`

**Reply by p4block** [reply]
I just added it manually from sd-boot and the problem persists, seems like that doesn't avoid the problem. I'm going to just set cachefile=none on my pools again :sweat: 

Maybe somebody else has better luck?
